cpp/include/cudf/detail/scatter.cuh

```diff
@@ -0,0 +1,193 @@
+{
+  static_assert(std::is_signed<T>::value,
+    "Need different invalid index if unsigned index types are added");
+  auto const invalid_index = static_cast<T>(-1);
+
+  // Convert scatter map to a gather map
+  auto gather_map = rmm::device_vector<T>(gather_rows, invalid_index);
+  thrust::scatter(rmm::exec_policy(stream)->on(stream),
+    thrust::make_counting_iterator<T>(0),
+    thrust::make_counting_iterator<T>(std::distance(scatter_map_begin, scatter_map_end)),
+    scatter_map_begin,
+    gather_map.begin());
+
+  return gather_map;
+}
+
+template <typename MapIterator>
+void gather_bitmask(table_view const& source, MapIterator gather_map,
+    std::vector<std::unique_ptr<column>>& target,
+    rmm::mr::device_memory_resource* mr, cudaStream_t stream)
+{
+  // Create null mask if source is nullable but target is not
+  for (size_t i = 0; i < target.size(); ++i) {
+    if (source.column(i).nullable() and not target[i]->nullable()) {
+      auto mask = create_null_mask(target.size(), mask_state::ALL_VALID, stream, mr);
+      target[i]->set_null_mask(std::move(mask), 0);
+    }
+  }
+
+  auto const device_source = table_device_view::create(source, stream);
+
+  // Make device array of target bitmask pointers
+  thrust::host_vector<bitmask_type*> target_masks(target.size());
+  std::transform(target.begin(), target.end(), target_masks.begin(),
+    [](auto const& col) { return col->mutable_view().null_mask(); });
+  rmm::device_vector<bitmask_type*> d_target_masks(target_masks);
+  auto target_rows = target.front()->size();
+
+  // Compute block size
+  auto bitmask_kernel = gather_bitmask_kernel<true, decltype(gather_map)>;
+  constexpr size_type block_size = 256;
+  size_type const grid_size = grid_1d(target_rows, block_size).num_blocks;
+
+  auto d_valid_counts = rmm::device_vector<size_type>(target.size());
+  bitmask_kernel<<<grid_size, block_size, 0, stream>>>(*device_source,
+    gather_map, d_target_masks.data().get(), target_rows, d_valid_counts.data().get());
+
+  // Copy the valid counts into each column
+  auto const valid_counts = thrust::host_vector<size_type>(d_valid_counts);
+  size_t index = 0;
+  for (auto& target_col : target) {
+    if (target_col->nullable()) {
+      auto const null_count = target_rows - valid_counts[index++];
+      target_col->set_null_count(null_count);
+    }
+  }
+}
+
+template <typename MapIterator>
+struct column_scatterer {
+  template <typename T, std::enable_if_t<is_fixed_width<T>()>* = nullptr>
+  std::unique_ptr<column> operator()(column_view const& source,
+      MapIterator scatter_map_begin, MapIterator scatter_map_end, column_view const& target,
+      rmm::mr::device_memory_resource* mr, cudaStream_t stream)
+  {
+    auto result = std::make_unique<column>(target, stream, mr);
+    auto result_view = result->mutable_view();
+
+    // NOTE use source.begin + scatter rows rather than source.end in case the
+    // scatter map is smaller than the number of source rows
```

Comment:
This comment is the opposite of the behaviour immediately following it. This indicates the source size is limiting whereas the code says iterator distance is limiting.

Reply:
```
thrust::scatter(rmm::exec_policy(stream)->on(stream), source.begin<T>(),
      source.begin<T>() + std::distance(scatter_map_begin, scatter_map_end),
      scatter_map_begin,
      result_view.begin<T>());
```
I am limiting the source size to size of the scatter map. Do you want to have two conditions where one limited by source size if source.size() < map.size() and other one which already exists?

Reply:
I misunderstood something.